### ISSUE TITLE
highgui: imshow can only show CV_8U images

### ISSUE THREAD

**Issue by berak** [opencv/opencv#13216]
- OpenCV => master
- Operating System / Platform => Windows 64 Bit
- Compiler => mingw64 / 7.2.0

##### Detailed description
regression after: https://github.com/opencv/opencv/commit/11eafca3e2a4cbc62f1309d25db0ea3ed9a6ea8e

the `cvConvertImage()` calls [were removed](https://github.com/opencv/opencv/commit/11eafca3e2a4cbc62f1309d25db0ea3ed9a6ea8e#diff-baec79d9f6cf2a8b605a5d9aad326540L1214). 
while code was added to adapt the colorspace, there is no more depth conversion to CV_8U

##### Steps to reproduce
```
//Mat m(400,400,CV_32FC3,Scalar(1.0));
Mat m(400,400,CV_16U,Scalar(255));
imshow("win",m);
waitKey();
```
leading to:

```
OpenCV(4.0.0-pre) Error: Assertion failed (dst.data == (uchar*)dst_ptr) in cvShowImage, file C:\p\opencv\modules\highgui\src\window_w32.cpp, line 1230
```
apart from the windows version (tried here), others will be affected: 
- [X] window_w32.cpp 
- [ ] window_cocoa.mm 
- [ ] window_gtk.cpp
- [x] window_QT.cpp

**Comment by mshabunin** [comment]
Qt backend has been fixed: #13184

**Comment by tompollok** [comment]
I faced the same issue while testing OpenCV 4.0.0 gold
https://github.com/opencv/opencv/blob/master/samples/cpp/dis_opticalflow.cpp

OpenCV(4.0.0) C:\build\master_winpack-build-win64-vc14\opencv\modules\highgui\src\window_w32.cpp:1230: error: (-215:Assertion failed) dst.data == (uchar*)dst_ptr in function 'cvShowImage'
